src/ray/gcs/gcs_server/gcs_task_manager.cc

```diff
@@ -30,7 +30,24 @@ void GcsTaskManager::Stop() {
 
 std::vector<rpc::TaskEvents> GcsTaskManager::GcsTaskManagerStorage::GetTaskEvents()
     const {
-  return task_events_;
+  std::vector<rpc::TaskEvents> ret;
+  // NOTE(rickyx): This could be done better if we expose an iterator - which we
+  // probably have to do if we are supporting pagination in the future.
+  // As for now, this will make sure data is returned w.r.t insertion order, so we could
+  // return the more recent entries when limit applies.
+  RAY_CHECK(next_idx_to_overwrite_ < task_events_.size())
+      << "next_idx_to_overwrite should be in bound.";
+  // Copy from the next insertion index
+  std::copy(task_events_.begin() + next_idx_to_overwrite_,
+            task_events_.end(),
+            std::back_inserter(ret));
+  // Copy the wrapped around if any
```

Comment:
Yes - and I am iterating from the end when processing this return. This avoids an explicit reverse.

Reply:
Hmm actually I am a bit confused here. If we insert least recently inserted items first, wouldn't that mean when you call `ray list tasks --limit 100`, it will not return the most recent tasks?

Reply:
There is also the change to iterate from the end here tho: https://github.com/ray-project/ray/pull/31837/files#diff-bb7747741306d767d5a35c0e55cde709d54cc407b24c806a61d63392795e1303R338

Reply:
But that's "after" we return the 100 lease recently inserted values? 

So, if you have

1, 2, 3, 4, 5

and then if the limit is 3

we will return 3, 2, 1.

But we should return 5, 4, 3 instead?

Reply:
We will return 5,4 3 here when iterating on all the returned values. The limit (for truncation) is applied before returning to the client

Reply:
ah so this logic will just re-sort all events (and not applying any truncate)

Reply:
It does apply truncate on the most recently inserted values.